src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,926 @@
+                                {
+                                    monotonicity.direction *= child_match.monotonicity->direction;
+                                    if (!child_match.monotonicity->strict)
+                                        monotonicity.strict = false;
+                                }
+
+                                match.node = child_match.node;
+                                match.monotonicity = monotonicity;
+                            }
+                        }
+                    }
+                }
+            }
+
+            stack.pop();
+        }
+    }
+
+    return matches;
+}
+
+InputOrderInfoPtr buildInputOrderInfo(
+    const FixedColumns & fixed_columns,
+    const ActionsDAGPtr & dag,
+    const SortDescription & description,
+    const ActionsDAG & sorting_key_dag,
+    const Names & sorting_key_columns,
+    size_t limit)
+{
+    //std::cerr << "------- buildInputOrderInfo " << std::endl;
+    SortDescription order_key_prefix_descr;
+    order_key_prefix_descr.reserve(description.size());
+
+    MatchedTrees::Matches matches;
+    FixedColumns fixed_key_columns;
+
+    if (dag)
+    {
+        matches = matchTrees(sorting_key_dag, *dag);
+
+        for (const auto & [node, match] : matches)
+        {
+            //std::cerr << "------- matching " << static_cast<const void *>(node) << " " << node->result_name
+            //    << " to " << static_cast<const void *>(match.node) << " " << (match.node ? match.node->result_name : "") << std::endl;
+            if (!match.monotonicity || match.monotonicity->strict)
+            {
+                if (match.node && fixed_columns.contains(node))
+                    fixed_key_columns.insert(match.node);
+            }
+        }
+
+        enreachFixedColumns(sorting_key_dag, fixed_key_columns);
+    }
+
+    /// This is a result direction we will read from MergeTree
+    ///  1 - in order,
+    /// -1 - in reverse order,
+    ///  0 - usual read, don't apply optimization
+    ///
+    /// So far, 0 means any direction is possible. It is ok for constant prefix.
+    int read_direction = 0;
+    size_t next_descr_column = 0;
+    size_t next_sort_key = 0;
+
+    while (next_descr_column < description.size() && next_sort_key < sorting_key_columns.size())
+    {
+        const auto & sorting_key_column = sorting_key_columns[next_sort_key];
+        const auto & descr = description[next_descr_column];
+
+        /// If required order depend on collation, it cannot be matched with primary key order.
+        /// Because primary keys cannot have collations.
+        if (descr.collator)
+            break;
+
+        /// Direction for current sort key.
+        int current_direction = 0;
+        bool strict_monotonic = true;
+
+        const ActionsDAG::Node * sort_column_node = sorting_key_dag.tryFindInOutputs(sorting_key_column);
+        /// This should not happen.
+        if (!sort_column_node)
+            break;
+
+        if (!dag)
```

Comment:
Is it possible? Shouldn't we build dag just for input columns even without any functions?

Reply:
It is possible. This DAG is from SELECT. For example, `SELECT * FROM tab ORDER BY a, b` will not have any expression steps, and DAG will be empty.